### ISSUE TITLE
Error: [birpc] timeout on calling "fetch"

### ISSUE THREAD

**Issue by Maxim-Mazurok** [vitest-dev/vitest#5182]
### Describe the bug

Getting `Error: [birpc] timeout on calling "fetch"` repeatedly after upgrading

### Reproduction

https://github.com/Maxim-Mazurok/vite-repro

### System Info

```shell
System:
    OS: Windows 11 10.0.22621
    CPU: (20) x64 12th Gen Intel(R) Core(TM) i9-12900H
    Memory: 27.99 GB / 63.68 GB
  Binaries:
    Node: 20.10.0 - ~\AppData\Local\nvs\node\20.10.0\x64\node.EXE
    npm: 10.2.3 - ~\AppData\Local\nvs\node\20.10.0\x64\npm.CMD
  Browsers:
    Edge: Chromium (121.0.2277.83)
    Internet Explorer: 11.0.22621.1
  npmPackages:
    @vitest/coverage-v8: 1.0.4 => 1.0.4
    @vitest/ui: 1.0.4 => 1.0.4
    vite: >=5.1.1 => 5.1.1
    vitest: 1.0.4 => 1.0.4
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by Maxim-Mazurok** [comment]
This is happening for 2/20 nx projects I have.
And only happens on vite 5.1.x but works fine on 5.0.x

**Comment by ikar1234** [comment]
If you want a quick fix, and your setup allows it, you can try https://github.com/vitest-dev/vitest/issues/4106#issuecomment-1713274961, This worked for us.

**Comment by Maxim-Mazurok** [comment]
> If you want a quick fix, and your setup allows it, you can try [#4106 (comment)](https://github.com/vitest-dev/vitest/issues/4106#issuecomment-1713274961), This worked for us.

I'm getting timeout on high-end PC, not in CI/CD:
Processor	12th Gen Intel(R) Core(TM) i9-12900H   2.50 GHz
Installed RAM	64.0 GB (63.7 GB usable)

I've tried to increase timeout as suggested, but it doesn't help. 3/20 projects just hang on vite with zero activity, clearly not a performance issue, they aren't even trying to run as far as I can tell

**Comment by ak274** [comment]
I'm also facing this error while running tests in a nx project with vite@5.1.1 but working fine with vite@5.0.12

**Comment by pongells** [comment]
same error here with vite 5.1.1

**Comment by Maxim-Mazurok** [comment]
Some more reproduction help:

It doesn't repro when I only have 1 test/spec file in the project.

If I have 2 test/spec files (even if they are empty) - I get a stable reproduction.

I am running an individual project (`nx run utils:test`) not in parallel.

Here's my `vite.config.ts` for the project:
```ts
import { defineConfig } from "vite";
import { merge } from "ts-deepmerge";
import viteTsConfigPaths from "vite-tsconfig-paths";
// eslint-disable-next-line @nx/enforce-module-boundaries
import { myDefaultUserConfig } from "../../vite-stuff";

export default defineConfig(
  merge(myDefaultUserConfig, {
    cacheDir: "../../node_modules/.vite/math",

    plugins: [
      viteTsConfigPaths({
        root: "../../",
      }),
    ],

    test: {
      cache: {
        dir: "../../node_modules/.vitest",
      },
      environment: "node",
      include: ["src/**/*.{test,spec}.{js,mjs,cjs,ts,mts,cts,jsx,tsx}"],
    },
  }),
);
```

Here's my `vite-stuff.ts`:
```ts
import { fileURLToPath } from "node:url";
import viteTsConfigPaths from "vite-tsconfig-paths";
import { UserConfig } from "vitest/config";
// import { pathToResources } from "@nx/nx-utils"; // TODO: figure out how to re-arrange imports
//@ts-expect-error doesn't matter, seems to work fine
const pathToResources = new URL("resources/", import.meta.url);

interface MyUserConfig extends UserConfig {
  plugins: Required<UserConfig["plugins"]>;
  test: Required<
    Pick<
      Exclude<Required<UserConfig["test"]>, undefined>,
      "globals" | "restoreMocks" | "reporters"
    > & {
      setupFiles: Exclude<
        Pick<
          Exclude<Required<UserConfig["test"]>, undefined>,
          "setupFiles"
        >["setupFiles"],
        string
      >;
    }
  >;
}

export const myDefaultUserConfig: MyUserConfig = {
  plugins: [viteTsConfigPaths()],
  test: {
    globals: true,
    setupFiles: [
      fileURLToPath(new URL("setupFiles/toEqual.ts", pathToResources)), // requires server.fs.strict to be false to work
    ],
    restoreMocks: false, // TODO: enable this, but beware of issues with vi.mock tho (TypeError: Class constructor GoogleSheetsDownloader cannot be invoked without 'new')
    reporters: ["default"],
  },
  server: {
    fs: {
      strict: false, // this will allow to load setup file from resources
    },
  },
};
```

Hope this helps to resolve the issue, cheers!

**Comment by mwood23** [comment]
Same, two spec files and it breaks reliably.

**Comment by Maxim-Mazurok** [comment]
> Hello @Maxim-Mazurok. Please provide a [minimal reproduction](https://stackoverflow.com/help/minimal-reproducible-example) using a GitHub repository or [StackBlitz](https://vitest.new). Issues marked with `need reproduction` will be closed if they have no activity within 3 days.

Added reproduction on a clean nx setup, see readme for steps I took to set it up: https://github.com/Maxim-Mazurok/vite-repro

**Comment by sheremet-va** [comment]
This is an issue with how Vite resolves tsconfigs. I think running it with nx makes global Vite cache invalid, and so it hangs. Related:

- https://github.com/dominikg/tsconfck/issues/154

**Comment by sheremet-va** [comment]
By the way, the workaround is to remove references from tsconfig:

```diff
{
  "extends": "../tsconfig.base.json",
  "compilerOptions": {
    "module": "commonjs",
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    "noImplicitOverride": true,
    "noPropertyAccessFromIndexSignature": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true
  },
  "files": [],
  "include": [],
-  "references": [
-    {
-      "path": "./tsconfig.lib.json"
-    },
-    {
-      "path": "./tsconfig.spec.json"
-    }
-  ]
}

```

**Comment by daihere1993** [comment]
I'm facing the exactly same issue while using nx with vitest. I'm using newer vitest then the error message would be:
_"Error: [vitest-worker]: Timeout calling "fetch" with "["/xxxx/demo_repos/vite-repro/bla/src/lib/bla.spec.ts","ssr"]""_

which modified by https://github.com/vitest-dev/vitest/pull/5103

**Take https://github.com/Maxim-Mazurok/vite-repro as an example, the better W/A would be modify the `extends` of `tsconfig.spec.json` and `tsconfig.lib.json`  to `"extends": "../tsconfig.base.json"` from `"extends": "./tsconfig.json"`**
```diff
diff --git a/bla/tsconfig.lib.json b/bla/tsconfig.lib.json
index 2fd1e3f..4d0c527 100644
--- a/bla/tsconfig.lib.json
+++ b/bla/tsconfig.lib.json
@@ -1,5 +1,5 @@
 {
-  "extends": "./tsconfig.json",
+  "extends": "../tsconfig.base.json",
   "compilerOptions": {
     "outDir": "../dist/out-tsc",
     "declaration": true,
diff --git a/bla/tsconfig.spec.json b/bla/tsconfig.spec.json
index 4f7fed6..3e6ed46 100644
--- a/bla/tsconfig.spec.json
+++ b/bla/tsconfig.spec.json
@@ -1,5 +1,5 @@
 {
-  "extends": "./tsconfig.json",
+  "extends": "../tsconfig.base.json",
   "compilerOptions": {
     "outDir": "../dist/out-tsc",
     "types": [
```

Seems vitest has some issues to support `extends`. By the way, there is another issue https://github.com/vitest-dev/vitest/issues/3077 which threw same timeout error message and suggested W/A is to use `--pool=forks` I tried but still not working.

@Maxim-Mazurok @sheremet-va @AriPerkkio FYI.

**Comment by chinanderm** [comment]
@daihere1993 I am running into the same problem in my Nx repo with vite 5.1.x and vitest 1.3.x. In the workaround you suggested, that is the exact structure of the `tsconfig` files in our repo and the issue remains. You can further clarify this workaround?

**Comment by daihere1993** [comment]
> @daihere1993 I am running into the same problem in my Nx repo with vite 5.1.x and vitest 1.3.x. In the workaround you suggested, that is the exact structure of the `tsconfig` files in our repo and the issue remains. You can further clarify this workaround?

The issue we facing is a known bug(https://github.com/dominikg/tsconfck/issues/154) of [tsconfck](https://github.com/dominikg/tsconfck) which is a dependency package of vite to parse tsconfig files. 


```
tsconfig.base.json
/apps/{your_app}/tsconfig.json # references ['./tsconfig.app.json', 'tsconfig.spec.json']
/apps/{your_app}/tsconfig.app.json # extends './tsconfig.json'
/apps/{your_app}/tsconfig.spec.json # extends './tsconfig.json'
```
Above is a typical structure and typescript configuration of nx app with vitest. As you can see the `tsconfig.json` refer to `tsconfig.app.json` and `tsconfig.spec.json` then both of them extend back to `tsconfig.json`. The current implementation of  [tsconfck](https://github.com/dominikg/tsconfck) can not handle such "circle" configuration which will cause a deadlock situation in which referenced file and extended file would `await` the parsing of each other then timeout happened.

Thus the W/A is to avoid such "circle", and just extends `tsconfig.base.json` directly.

**Comment by viceice** [comment]
strange, for me only `--pool=forks` cli option works, setting it inside `vitest.config.js` has no effect.

**Comment by sapphi-red** [comment]
The fix on Vite side is released in 5.1.6 (https://github.com/vitejs/vite/pull/16124).
